src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +51,288 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
+
+    type FileResultsComplete =
+        {
+            Phase1: Phase1Res
+            Phase2: Phase2Res
+            Phase3: Phase3Res
+        }
+
+    type FilePhaseFuncs = Phase1Fun * Phase2Fun * Phase3Fun
+
+    [<RequireQualifiedAccess>]
+    type OptimizationPhase =
+        | Phase1
+        | Phase2
+        | Phase3
+
+    type FileResults =
+        {
+            mutable Phase1: Phase1Res option
+            mutable Phase2: Phase2Res option
+            mutable Phase3: Phase3Res option
+        }
+
+        member this.HasResult(phase: OptimizationPhase) =
+            match phase with
+            | OptimizationPhase.Phase1 -> this.Phase1 |> Option.isSome
+            | OptimizationPhase.Phase2 -> this.Phase2 |> Option.isSome
+            | OptimizationPhase.Phase3 -> this.Phase3 |> Option.isSome
+
+        static member Empty =
+            {
+                Phase1 = None
+                Phase2 = None
+                Phase3 = None
+            }
+
+    type Node =
+        {
+            Idx: int
+            Phase: OptimizationPhase
+        }
+
+        override this.ToString() = $"[{this.Idx}-{this.Phase}]"
+
+    let collectResults
+        (inputs: FileResultsComplete[])
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files =
+            inputs
+            |> Array.map
+                (fun {
+                         Phase1 = phase1
+                         Phase2 = _phase2
+                         Phase3 = phase3
+                     } ->
+                    let (_, _, implFileOptData, _), optimizeDuringCodeGen = phase1
+                    let _, implFile = phase3
+
+                    let implFile =
+                        {
+                            ImplFile = implFile
+                            OptimizeDuringCodeGen = optimizeDuringCodeGen
+                        }
+
+                    implFile, implFileOptData)
+
+        let lastFilePhase1Env =
+            inputs
+            |> Array.last
+            |> fun { Phase1 = phase1 } ->
+                let (optEnvPhase1, _, _, _), _ = phase1
+                optEnvPhase1
+
+        files, lastFilePhase1Env
+
+    let getPhase2Res (p: FileResults) = p.Phase2 |> Option.get
+
+    let getPhase3Res (p: FileResults) =
+        p.Phase3 |> Option.get |> (fun (env, _) -> env)
+
+    let optimizeFilesInParallel
+        (env0: IncrementalOptimizationEnv)
+        ((phase1, phase2, phase3): FilePhaseFuncs)
+        (files: CheckedImplFile list)
+        (ct: CancellationToken)
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files = files |> List.toArray
+
+        let firstNode =
+            {
+                Idx = 0
+                Phase = OptimizationPhase.Phase1
+            }
+
+        let results = files |> Array.map (fun _ -> FileResults.Empty)
+
+        let _lock = obj ()
```

Comment:
This `_lock` has me worried. If you are doing parallelism, you normally do not want to have a lock that happens every time. 

Can you describe what the lock is trying to do here?

Reply:
Uaing a lock isn't ideal, but the cost of a lock is small when no blocking occurs, and because the operation in the lock is quick, blocking should almost never occur and be very short if it does.

But definitely avoiding the lock would be nice.

I'll add more comments on why this is needed later.

Reply:
What is your main concern with the lock?

Reply:
I'm wondering why not use the existing "nodeCode" and build graph nodes?   Alternatively why not use "Task" based coding?